src/Sulu/Bundle/MediaBundle/Media/Storage/FlysystemStorage.php

```diff
@@ -0,0 +1,143 @@
+<?php
+/*
+ * This file is part of the Sulu CMS.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\Media\Storage;
+
+use League\Flysystem\FilesystemInterface;
+use League\Flysystem\MountManager;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Sulu\Bundle\MediaBundle\Media\Exception\FilenameAlreadyExistsException;
+use Sulu\Bundle\MediaBundle\Media\Exception\FilenameWriteException;
+use Sulu\Bundle\MediaBundle\Media\Storage\Resolver\FlysystemResolverInterface;
+
+/**
+ * Used Flysystem to store media files.
+ */
+class FlysystemStorage extends AbstractStorage
+{
+    const STORAGE_OPTION_FILENAME = 'fileName';
+    const STORAGE_OPTION_URL = 'url';
+    const STORAGE_TYPE = 'FlySystem';
+
+    /**
+     * @var string
+     */
+    protected $fileSystem;
+
+    /**
+     * @var FlysystemResolverInterface
+     */
+    protected $flysystemResolver;
+
+    /**
+     * @var NullLogger|LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * @param string $fileSystem
+     * @param MountManager $mountManager
+     * @param FlysystemResolverInterface $flysystemResolver
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        $fileSystem,
+        MountManager $mountManager,
+        FlysystemResolverInterface $flysystemResolver,
+        $logger = null
+    ) {
+        $this->fileSystem = $fileSystem;
+        $this->mountManager = $mountManager;
+        $this->flysystemResolver = $flysystemResolver;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save($tempPath, $fileName, $storageOptions = null)
+    {
+        $this->storageOptions = new \stdClass();
```

Comment:
IIRC the storage should be a singleton service instance, so it seems a bit odd, that storage class holds a state of the last file saved/loaded..